File: python/slhdsa/wots.py
"""WOTS+ one-time signature for SLH-DSA per FIPS 205, Section 5."""

from slhdsa.params import SLHDSAParams
from slhdsa.address import ADRS, WOTS_HASH, WOTS_PK, WOTS_PRF
from slhdsa.hash_funcs import get_suite
from slhdsa.utils import base_2b, to_byte


def chain(params: SLHDSAParams, x: bytes, i: int, s: int,
          pk_seed: bytes, adrs: ADRS) -> bytes:
    """Chaining function (Algorithm 4).

    Applies F iteratively s times starting from step i.
    """
    if s == 0:
        return x
    if (i + s) > (params.w - 1):
        return None

    suite = get_suite(params.hash_type)
    tmp = chain(params, x, i, s - 1, pk_seed, adrs)
    adrs.set_hash_address(i + s - 1)
    tmp = suite.F(pk_seed, adrs, tmp, params.n)
    return tmp


def chain_iterative(params: SLHDSAParams, x: bytes, i: int, s: int,
                    pk_seed: bytes, adrs: ADRS) -> bytes:
    """Iterative version of chain to avoid deep recursion."""
    if s == 0:
        return x
    if (i + s) > (params.w - 1):
        return None

    suite = get_suite(params.hash_type)
    tmp = x
    for j in range(i, i + s):
        adrs.set_hash_address(j)
        tmp = suite.F(pk_seed, adrs, tmp, params.n)
    return tmp


def wots_pkgen(params: SLHDSAParams, sk_seed: bytes, pk_seed: bytes,
               adrs: ADRS) -> bytes:
    """WOTS+ public key generation (Algorithm 5)."""
    suite = get_suite(params.hash_type)
    wots_pk_adrs = adrs.copy()

    tmp = b""
    for i in range(params.wots_len):
        # Generate secret key element
        sk_adrs = adrs.copy()
        sk_adrs.set_type(WOTS_PRF)
        sk_adrs.set_key_pair_address(adrs.get_key_pair_address())
        sk_adrs.set_chain_address(i)
        sk_adrs.set_hash_address(0)
        sk = suite.PRF(pk_seed, sk_seed, sk_adrs, params.n)

        # Chain to get public key element
        wots_chain_adrs = adrs.copy()
        wots_chain_adrs.set_type(WOTS_HASH)
        wots_chain_adrs.set_key_pair_address(adrs.get_key_pair_address())
        wots_chain_adrs.set_chain_address(i)
        tmp += chain_iterative(params, sk, 0, params.w - 1, pk_seed, wots_chain_adrs)

    # Compress public key
    wots_pk_adrs.set_type(WOTS_PK)
    wots_pk_adrs.set_key_pair_address(adrs.get_key_pair_address())
    pk = suite.T(pk_seed, wots_pk_adrs, tmp, params.n)
    return pk


def wots_sign(params: SLHDSAParams, m: bytes, sk_seed: bytes,
              pk_seed: bytes, adrs: ADRS) -> bytes:
    """WOTS+ signature generation (Algorithm 6).

    Args:
        m: n-byte message digest to sign.
    """
    suite = get_suite(params.hash_type)

    # Convert message to base-w representation
    msg = base_2b(m, params.lg_w, params.wots_len1)

    # Compute checksum
    csum = 0
    for v in msg:
        csum += params.w - 1 - v

    # Pad checksum to full bytes and convert to base-w
    csum_bits = params.wots_len2 * params.lg_w
    csum_bytes = (csum_bits + 7) // 8
    csum_b = to_byte(csum << (8 * csum_bytes - csum_bits), csum_bytes)
    msg_csum = msg + base_2b(csum_b, params.lg_w, params.wots_len2)

    sig = b""
    for i in range(params.wots_len):
        sk_adrs = adrs.copy()
        sk_adrs.set_type(WOTS_PRF)
        sk_adrs.set_key_pair_address(adrs.get_key_pair_address())
        sk_adrs.set_chain_address(i)
        sk_adrs.set_hash_address(0)
        sk = suite.PRF(pk_seed, sk_seed, sk_adrs, params.n)

        chain_adrs = adrs.copy()
        chain_adrs.set_type(WOTS_HASH)
        chain_adrs.set_key_pair_address(adrs.get_key_pair_address())
        chain_adrs.set_chain_address(i)
        sig += chain_iterative(params, sk, 0, msg_csum[i], pk_seed, chain_adrs)

    return sig


def wots_pk_from_sig(params: SLHDSAParams, sig: bytes, m: bytes,
                     pk_seed: bytes, adrs: ADRS) -> bytes:
    """Compute WOTS+ public key from signature (Algorithm 7)."""
    suite = get_suite(params.hash_type)

    # Convert message to base-w
    msg = base_2b(m, params.lg_w, params.wots_len1)

    csum = 0
    for v in msg:
        csum += params.w - 1 - v

    csum_bits = params.wots_len2 * params.lg_w
    csum_bytes = (csum_bits + 7) // 8
    csum_b = to_byte(csum << (8 * csum_bytes - csum_bits), csum_bytes)
    msg_csum = msg + base_2b(csum_b, params.lg_w, params.wots_len2)

    tmp = b""
    for i in range(params.wots_len):
        chain_adrs = adrs.copy()
        chain_adrs.set_type(WOTS_HASH)
        chain_adrs.set_key_pair_address(adrs.get_key_pair_address())
        chain_adrs.set_chain_address(i)
        sig_i = sig[i * params.n:(i + 1) * params.n]
        tmp += chain_iterative(params, sig_i, msg_csum[i],
                               params.w - 1 - msg_csum[i], pk_seed, chain_adrs)

    wots_pk_adrs = adrs.copy()
    wots_pk_adrs.set_type(WOTS_PK)
    wots_pk_adrs.set_key_pair_address(adrs.get_key_pair_address())
    pk = suite.T(pk_seed, wots_pk_adrs, tmp, params.n)
    return pk

File: js/src/slhdsa/wots.js
/**
 * WOTS+ (Winternitz One-Time Signature) for SLH-DSA (FIPS 205)
 * Algorithms 5-8
 */

import { ADRS, ADDR_TYPE } from './address.js';
import { base2b, concat, toByte } from './utils.js';

/**
 * chain(X, i, s, pkSeed, adrs, hashSuite) - Algorithm 5
 * Apply F s times starting from index i
 */
function chain(X, i, s, pkSeed, adrs, params, hashSuite) {
  if (s === 0) return new Uint8Array(X);
  let tmp = new Uint8Array(X);
  for (let j = i; j < i + s; j++) {
    adrs.setHashAddress(j);
    tmp = hashSuite.F(pkSeed, adrs, tmp);
  }
  return tmp;
}

/**
 * wotsPkgen(skSeed, pkSeed, adrs, params, hashSuite) - Algorithm 6
 * Generate a WOTS+ public key
 */
function wotsPkgen(skSeed, pkSeed, adrs, params, hashSuite) {
  const { n, len, w } = params;
  const skAdrs = adrs.copy();
  skAdrs.setType(ADDR_TYPE.WOTS_PRF);
  skAdrs.setKeyPairAddress(adrs.getKeyPairAddress());

  const tmp = new Uint8Array(len * n);
  for (let i = 0; i < len; i++) {
    skAdrs.setChainAddress(i);
    const sk = hashSuite.PRF(pkSeed, skSeed, skAdrs);
    adrs.setChainAddress(i);
    const pk_i = chain(sk, 0, w - 1, pkSeed, adrs, params, hashSuite);
    tmp.set(pk_i, i * n);
  }

  const wotsPkAdrs = adrs.copy();
  wotsPkAdrs.setType(ADDR_TYPE.WOTS_PK);
  wotsPkAdrs.setKeyPairAddress(adrs.getKeyPairAddress());
  return hashSuite.Tl(pkSeed, wotsPkAdrs, tmp);
}

/**
 * wotsSign(M, skSeed, pkSeed, adrs, params, hashSuite) - Algorithm 7
 * Generate a WOTS+ signature for message M
 */
function wotsSign(M, skSeed, pkSeed, adrs, params, hashSuite) {
  const { n, len, w, lgw } = params;

  // Per FIPS 205: len1 = ceil(8n / lgw), len2 = len - len1
  const len1 = Math.ceil((8 * n) / lgw);
  const len2 = len - len1;

  const msg1 = base2b(M, lgw, len1);

  // Compute checksum
  let csum = 0;
  for (let i = 0; i < len1; i++) {
    csum += w - 1 - msg1[i];
  }

  // csum needs ceil(floor(log2(len1*(w-1)))+1) bits, packed into bytes, then base2b
  const csumBits = Math.floor(Math.log2(len1 * (w - 1))) + 1;
  const csumBytes = Math.ceil(csumBits / 8);
  csum <<= (8 * csumBytes - csumBits); // left-shift per spec: csum << (8 - (csumBits % 8)) % 8
  const csumBuf = toByte(csum, csumBytes);
  const msg2 = base2b(csumBuf, lgw, len2);

  const msgAll = msg1.concat(msg2);

  // Sign
  const skAdrs = adrs.copy();
  skAdrs.setType(ADDR_TYPE.WOTS_PRF);
  skAdrs.setKeyPairAddress(adrs.getKeyPairAddress());

  const sig = new Uint8Array(len * n);
  for (let i = 0; i < len; i++) {
    skAdrs.setChainAddress(i);
    const sk = hashSuite.PRF(pkSeed, skSeed, skAdrs);
    adrs.setChainAddress(i);
    const s = chain(sk, 0, msgAll[i], pkSeed, adrs, params, hashSuite);
    sig.set(s, i * n);
  }
  return sig;
}

/**
 * wotsPkFromSig(sig, M, pkSeed, adrs, params, hashSuite) - Algorithm 8
 * Compute WOTS+ public key from signature
 */
function wotsPkFromSig(sig, M, pkSeed, adrs, params, hashSuite) {
  const { n, len, w, lgw } = params;

  const len1 = Math.ceil((8 * n) / lgw);
  const len2 = len - len1;

  const msg1 = base2b(M, lgw, len1);

  let csum = 0;
  for (let i = 0; i < len1; i++) {
    csum += w - 1 - msg1[i];
  }

  const csumBits = Math.floor(Math.log2(len1 * (w - 1))) + 1;
  const csumBytes = Math.ceil(csumBits / 8);
  csum <<= (8 * csumBytes - csumBits);
  const csumBuf = toByte(csum, csumBytes);
  const msg2 = base2b(csumBuf, lgw, len2);

  const msgAll = msg1.concat(msg2);

  const tmp = new Uint8Array(len * n);
  for (let i = 0; i < len; i++) {
    adrs.setChainAddress(i);
    const sigBlock = sig.slice(i * n, (i + 1) * n);
    const pk_i = chain(sigBlock, msgAll[i], w - 1 - msgAll[i], pkSeed, adrs, params, hashSuite);
    tmp.set(pk_i, i * n);
  }

  const wotsPkAdrs = adrs.copy();
  wotsPkAdrs.setType(ADDR_TYPE.WOTS_PK);
  wotsPkAdrs.setKeyPairAddress(adrs.getKeyPairAddress());
  return hashSuite.Tl(pkSeed, wotsPkAdrs, tmp);
}

export { chain, wotsPkgen, wotsSign, wotsPkFromSig };
